### ISSUE TITLE
Regression: ParamSpecs are converted to '...' when used in returned Callable

### ISSUE THREAD

**Issue by GugelRobin** [microsoft/pyright#5283]
**Describe the bug**
See code.

Pyright >1.1.309 starts having the bug

Pylance has the problem as well.

**Code or Screenshots**

```python

from typing import TypeVar, ParamSpec

T = TypeVar("T")
P = ParamSpec("P")

def foo() -> Callable[[Callable[P, T]], Callable[P, T]]:
    ...

def bar(x: int) -> str:
    ...

reveal_type(foo()(bar)) 
# pyright >  1.1.310 outputs: Type of "foo()(bar)" is "(...) -> str"
# pyright <= 1.1.309 outputs: Type of "foo()(bar)" is "(x: int) -> str"


```

**Comment by erictraut** [comment]
This is an ambiguous part of the typing spec. In a strict reading of PEP 484, both `P` and `T` are scoped to `foo` and must be solved in the context of a call to `foo`. However, pyright and mypy provide some special-case logic when a type variable is used only within a return callable where they assume that the type variable is scoped to the callable rather than the function. It looks like I broke that special casing for ParamSpec but not for regular type variables. I'll investigate.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.314, which I just published. It will also be included in this week's Pylance prerelease build.

**Comment by rslinckx** [comment]
@erictraut I have the following test case which seems to also have ceased working after 1.1.309, and seems related:

```python
from __future__ import annotations

from dataclasses import dataclass
from typing import Callable, Generic, ParamSpec

P = ParamSpec("P")


@dataclass
class TaskDeclaration(Generic[P]):
    func: Callable[P, None]

    def enqueue(self, something: str, *args: P.args, **kwargs: P.kwargs) -> None:
        pass


def task() -> Callable[[Callable[P, None]], TaskDeclaration[P]]:
    def decorator(func: Callable[P, None]) -> TaskDeclaration[P]:
        return TaskDeclaration(func)

    return decorator


@task()
def test(a: int) -> None:
    pass


test.enqueue("something") # Should have been test.enqueue("something", 3) to be valid
```

Output with <=1.1.309: test.py:29:1 - error: Argument missing for parameter "a" (reportGeneralTypeIssues)
Output with >1.1.309(including 314): 0 errors, 0 warnings, 0 informations

**Comment by erictraut** [comment]
Ah, my bad. I missed a small detail that caused my previous fix not to work with a class parameterized with a ParamSpec, as with `TaskDeclaration` in your example above. I've fixed the issue and further augmented the pyright unit tests to cover this case.

**Comment by Hootner** [comment]
Python
from typing import Callable, TypeVar, ParamSpec

T = TypeVar("T")
P = ParamSpec("P")

def foo() -> Callable[[Callable[P, T]], Callable[P, T]]:
    ...

def bar(x: int) -> str:
    ...

# Checking the type of "foo()(bar)"
from typing import reveal_type
reveal_type(foo()(bar))
# In pyright <= 1.1.309: Type of "foo()(bar)" is "(x: int) -> str"
# In pyright >  1.1.310: Type of "foo()(bar)" is "(...) -> str"
The issue here is that in Pyright version 1.1.310 and later (as well as in Pylance), there's a change in behavior when handling ParamSpec. In earlier versions (1.1.309 and below), Pyright correctly identifies the type of "foo()(bar)" as "(x: int) -> str". However, in later versions (1.1.310 and above), it shows the type as "(...) -> str", which is less specific.

**Comment by erictraut** [comment]
This is included in pyright 1.1.315, which I just published. It will also be included in this week's insiders release of pylance.
